keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,858 @@
+- Test the end-to-end workflow of updating `CSINode.Spec.Drivers[*].Allocatable.Count` using a CSI driver.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag.
+- Initial unit tests/integration tests completed and enabled.
+
+#### Beta
+
+- Allowing time for feedback (at least 2 releases between beta and GA).
+- All unit tests/integration/e2e tests completed and enabled.
+  - Validate kubelet behavior when API server rejects `CSINode` updates (older API server version).
+  - Validate CSI driver behavior with and without the `AllocatableUpdateInterval` field set.
+  - Validate scheduler behavior remains consistent regardless of whether `CSINode.Spec.Drivers[*].Allocatable.Count` is being dynamically updated or not.
+  - Validate pod construction failure handling in kubelet, ensuring it correctly updates the `CSINode` object when the feature is enabled and the API server supports it.
+
+#### GA
+
+- All beta criteria have been satisfied.
+- Feature is stable.
+- No bug reports / feedback / improvements to address.
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+- Upgrade Strategy
+  - Upgrade the API server first to support mutable `CSINode.Spec.Drivers[*].Allocatable.Count` and the new `AllocatableUpdateInterval` field in `CSIDriver` object.
+  - Upgrade nodes
+  - Update CSI drivers to take advantage of the new feature, if desired.
+
+- Downgrade Strategy
+  - If downgrading the API server, ensure that nodes are downgraded first to avoid rejected `CSINode` update attempts.
+  - CSI drivers using the `AllocatableUpdateInterval` feature should be reconfigured to not use this field before downgrading the API server.
+
+### Version Skew Strategy
+
+This enhancement primarily involves changes to the kubelet and the API server, with no impact on the scheduler. Here's how the system will behave in various version skew scenarios:
+
+- API Server considerations
+  - Older API server versions will reject updates to the `CSINode.Spec.Drivers[*].Allocatable.Count` field and won't recognize the `AllocatableUpdateInterval` field in the `CSIDriver` object.
+
+- Kubelet version considerations
+  - Newer kubelet (with this feature) + Older API server: The kubelet will attempt to update the `CSINode.Spec.Drivers[*].Allocatable.Count` field due to capacity failures, but these updates will be rejected by the API server.
```

Comment:
with the `AllocatableUpdateInterval` field not present (unknown fields are pruned on read), how often does the kubelet update the count?

Reply:
If `AllocatableUpdateInterval` is not explicitly defined, then Kubelet will only update the `CSINode.Spec.Drivers[*].Allocatable.Count` field when a pod fails to enter the running state because a volume in the pod spec failed to attach due to no capacity on the node. This KEP deems this an invalid state (scheduler should _never_ assign a stateful pod to a node that does not have capacity to support the attachment). 

As for how often this invalid state is reached, it depends on external factors such as a cluster operator manually attaching volumes outside of Kubernetes, or CNI plugins dynamically attaching network interfaces as the user scales up their workload (done if number of available `IPv4` addresses are insufficient for pod networking) -- these are operations that typically happen infrequently.

We also discussed the idea of building a protection mechanism to prevent more than X number of updates within X number of minutes, which I think is a good suggestion!